core/src/main/scala/org/http4s/EntityDecoder.scala

```diff
@@ -106,11 +106,19 @@ object EntityDecoder extends EntityDecoderInstances {
       msg.headers.get(`Content-Type`) match {
         case Some(contentType) =>
           if (a.matchesMediaType(contentType.mediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMismatch(actual, expected) =>
+              MediaTypeMismatch(actual, expected ++ a.consumes)
+            case other => other
+          }
 
         case None =>
           if (a.matchesMediaType(UndefinedMediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMissing(expected) =>
+              MediaTypeMissing(expected ++ a.consumes)
```

Comment:
Actually, although I have somewhat convinced myself it might be better for the `matchesMediaType` method to accept an `Option[MediaRange]` instead of simply a `MediaRange` (removing the need for `UndefinedMediaType`), such a change would not allow us to restructure things in a way that we could prove that we only will receive one type of error in each `case` so I guess we will just have to trust ourselves that the implementation is bug-free or handle both error types in both cases. Sorry I know I am kind of splitting hairs here, we are not programming in Idris here after all!